quinn/src/connection.rs

```diff
@@ -863,47 +876,56 @@ pub(crate) struct State {
     ref_count: usize,
     socket: Box<dyn AsyncUdpSocket>,
     runtime: Arc<dyn Runtime>,
+    /// We buffer a transmit when the underlying `poll_send` can't complete
+    buffered_transmit: Option<udp::Transmit>,
 }
 
 impl State {
-    fn drive_transmit(&mut self) -> bool {
+    fn drive_transmit(&mut self, cx: &mut Context) -> io::Result<bool> {
         let now = Instant::now();
         let mut transmits = 0;
 
         let max_datagrams = self.socket.max_transmit_segments();
         let capacity = self.inner.current_mtu();
         let mut buffer = BytesMut::with_capacity(capacity as usize);
 
-        while let Some(t) = self.inner.poll_transmit(now, max_datagrams, &mut buffer) {
-            transmits += match t.segment_size {
+        while let Some(t) = self.buffered_transmit.take().or_else(|| {
+            self.inner
+                .poll_transmit(now, max_datagrams, &mut buffer)
+                .map(|t| {
+                    let buffer = buffer.split_to(t.size).freeze();
+                    udp_transmit(t, buffer)
+                })
+        }) {
+            let size = t.contents.len();
+            let segment_size = t.segment_size;
+            if let Poll::Ready(0) | Poll::Pending =
+                self.socket.poll_send(cx, std::slice::from_ref(&t))?
+            {
+                self.buffered_transmit = Some(t);
+                return Ok(false);
+            }
+            transmits += match segment_size {
```

Comment:
The diff here gets a bit confusing because I think you inlined this refactoring of moving around the `transmits` calculation? I think it's a good change, but if it can be a separate commit it would be good to do that.

Reply:
Reduced the churn here. I don't think it actually matters much whether we count after successful `poll_transmit` or after `poll_send`, and the original pattern is a bit simpler.